golem/utils/asyncio/buffer.py

```diff
@@ -0,0 +1,403 @@
+        is empty.
+        """
+
+    @abstractmethod
+    async def put(self, item: TItem, *, lock=True) -> None:
+        """Add item to right-most position to buffer.
+
+        Duplicates are supported.
+        """
+
+    @abstractmethod
+    async def put_all(self, items: Sequence[TItem], *, lock=True) -> None:
+        """Replace all items stored in buffer.
+
+        Duplicates are supported.
+        """
+
+    @abstractmethod
+    async def remove(self, item: TItem, *, lock=True) -> None:
+        """Remove first occurrence of item from buffer or raise `ValueError` if not found."""
+
+    @abstractmethod
+    async def set_exception(self, exc: BaseException, *, lock=True) -> None:
+        """Set exception that will be raised while trying to `.get()`/`.get_all()` item from \
+        empty buffer."""
+
+    def reset_exception(self) -> None:
+        """Reset exception that was previously set by calling `.set_exception()`."""
+
+
+class ComposableBuffer(Buffer[TItem]):
+    """Utility class for composable/stackable buffer implementations to help with calling \
+    underlying buffer."""
+
+    def __init__(self, buffer: Buffer[TItem]):
+        self._buffer = buffer
+
+    @asynccontextmanager
+    async def _handle_lock(self, lock: bool):
+        if lock:
+            async with self._buffer.condition:
+                yield
+        else:
+            yield
+
+    def size(self) -> int:
+        return self._buffer.size()
+
+    async def wait_for_any_items(self, *, lock=True) -> None:
+        await self._buffer.wait_for_any_items(lock=lock)
+
+    async def get(self, *, lock=True) -> TItem:
+        return await self._buffer.get(lock=lock)
+
+    async def get_all(self, *, lock=True) -> MutableSequence[TItem]:
+        return await self._buffer.get_all(lock=lock)
+
+    async def put(self, item: TItem, *, lock=True) -> None:
+        await self._buffer.put(item, lock=lock)
+
+    async def put_all(self, items: Sequence[TItem], *, lock=True) -> None:
+        await self._buffer.put_all(items, lock=lock)
+
+    async def remove(self, item: TItem, *, lock=True) -> None:
+        await self._buffer.remove(item, lock=lock)
+
+    async def set_exception(self, exc: BaseException, *, lock=True) -> None:
+        await self._buffer.set_exception(exc, lock=lock)
+
+    def reset_exception(self) -> None:
+        self._buffer.reset_exception()
+
+
+class SimpleBuffer(Buffer[TItem]):
+    """Most basic implementation of Buffer interface."""
+
+    def __init__(self, items: Optional[Sequence[TItem]] = None):
+        self._items = list(items) if items is not None else []
+        self._error: Optional[BaseException] = None
+
+        self.condition = asyncio.Condition()
+
+    def size(self) -> int:
+        return len(self._items)
+
+    @asynccontextmanager
+    async def _handle_lock(self, lock: bool):
+        if lock:
+            async with self.condition:
+                yield
+        else:
+            yield
+
+    @trace_span()
```

Comment:
I don't really feel why we need to have those decorators on such low level objects... it's as if we wanted to have such logging on e.g. asyncio.Queue objects... 

plus, they're used inconsistenly, as far as I see...